src/backend/distributed/commands/utility_hook.c

```diff
@@ -912,6 +912,22 @@ ExecuteDistributedDDLJob(DDLJob *ddlJob)
 		{
 			CommitTransactionCommand();
 			StartTransactionCommand();
+
+			/*
+			 * Even if we started a new transaction, now we still hold
+			 * ExclusiveLock (acquired by VirtualXactLockTableInsert) on new
+			 * vxid and it is not released yet.
+			 *
+			 * Moreover, remote connections to localhost (that are opened by
+			 * Citus for shard level index commands) might enter into dead-lock
+			 * since they would attempt to acquire ShareLock on the same vxid
+			 * since postgres calls WaitForOlderSnapshots function when executing
+			 * CREATE INDEX (or REINDEX) CONCURRENTLY commands.
+			 *
+			 * To prevent self-deadlock, we release lock on vxid by explicitly
+			 * calling VirtualXactLockTableCleanup.
+			 */
+			VirtualXactLockTableCleanup();
```

Comment:
I think we probably cannot leave the transaction without virtual txId. Should we insert a new virtual tx ID after `ExecuteUtilityTaskList `?
(and, also what happens if the transaction fails without having the virt. tx Id)?

Reply:
>(and, also what happens if the transaction fails without having the virt. tx Id)?

I've no idea but I think many bad things might happen since we remove that lock as if xact finished, but it isn't actually 

>I think we probably cannot leave the transaction without virtual txId. Should we insert a new virtual tx ID after ExecuteUtilityTaskList ?

yeah maybe we should do:
```c
VirtualTransactionId vxid;
vxid.backendId = MyBackendId;
vxid.localTransactionId = GetNextLocalTransactionId();
VirtualXactLockTableInsert(vxid);
```

Reply:
I'm taking a brief look at this, and it doesn't look right to me but I think I'd need more context.

``VirtualXactLockTableCleanup()`` is only called from two places, and it doesn't seem like something we should be calling here. As @onderkalaci said, I'm concerned that we're releasing the vxid lock before the transaction completes.

Re-assigning a new virtual xact ID also seems quite strange to me. What about the time between when the first lock is released and the second one acquired? Even if this works, it feels like it adds complexity that will be hard to reason about in combination with other nuances of the locking system.